### PR TITLE
 Add varchar to char coercion for hive tables

### PR DESCRIPTION
**Pull request by Praveen2112**
## Description

Add varchar to char coercion for hive tables.

Enable the Hive users to retrieve the information from their tables after performing queries which change varchar/ string columns to float.

Hive sample DDL query:

ALTER TABLE mytable CHANGE COLUMN mycolumn mycolumn char(4);

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
# Hive
* Add varchar to char coercion for hive tables. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java [hunk 1]
```diff
@@ -131,6 +131,9 @@ public static Type createTypeFromCoercer(TypeManager typeManager, HiveType fromH
         if (fromType instanceof BooleanType && toType instanceof VarcharType toVarcharType) {
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java:131]: Document in `hive.md` https://trino.io/docs/current/connector/hive.html#schema-evolution the new schema evolution scenario

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestVarcharToCharCoercer.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import static io.trino.spi.type.VarcharType.createVarcharType;
+import static io.trino.type.InternalTypeManager.TESTING_TYPE_MANAGER;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestVarcharToCharCoercer
+{
+    @Test
+    public void testVarcharToSmallerCharCoercions()
+    {
+        assertVarcharToCharCoercion("a ", createVarcharType(3), "a", createCharType(2));
```
Comment (findinpath) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestVarcharToCharCoercer.java:38]: add as well

```
assertVarcharToCharCoercion("a", createVarcharType(3), "a", createCharType(2));
```

Comment (findinpath) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestVarcharToCharCoercer.java:38]:
```
assertVarcharToCharCoercion("   ", createVarcharType(3), "", createCharType(2));
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/BaseTestHiveCoercion.java [hunk 3]
```diff
@@ -163,6 +163,9 @@ protected void doTestHiveCoercion(HiveTableDefinition tableDefinition)
                 "date_to_bounded_varchar",
                 "char_to_bigger_char",
                 "char_to_smaller_char",
+                "string_to_char",
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/BaseTestHiveCoercion.java:166]: can you pls add a field dedicated to trimming ? 
we need to ensure that Trino functionality is compatible with Hive

Comment (Praveen2112) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/BaseTestHiveCoercion.java:166]: Do we need it specific to `String` as well - We have implemented one for `varchar_to_smaller_char`

Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/BaseTestHiveCoercion.java:166]: I was thinking more to have a concrete example from "abc  "  varchar(5) towards "abc" char(5)

**Comment by Praveen2112** [comment]
@findinpath Thanks a lot for your review. AC

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/BaseTestHiveCoercion.java [hunk 4]
```diff
@@ -267,6 +270,9 @@ protected void insertTableRows(String tableName, String floatToDoubleType)
                         "  DATE '2000-04-13', " +
                         "  'abc', " +
                         "  'abc', " +
+                        "  'Bigger Value', " +
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/BaseTestHiveCoercion.java:273]:
```
0: jdbc:hive2://localhost:10000/default> create table t1 (string_to_char string) partitioned by (id bigint) stored as orc;
0: jdbc:hive2://localhost:10000/default> insert into t1 values ('Hi',1), ('Hi     ', 2), ('Hello', 3);
0: jdbc:hive2://localhost:10000/default> alter table t1 change column string_to_char string_to_char char(3);
0: jdbc:hive2://localhost:10000/default> select '"' || string_to_char || '"' from t1;
+--------+
|  _c0   |
+--------+
| "Hi"   |
| "Hi"   |
| "Hel"  |
+--------+
(edited)
```
```
trino:default> select '"' || string_to_char || '"'  from hive.default.t1;
 _col0 
-------
 "Hel" 
 "Hi " 
 "Hi " 
(3 rows)
```

It seems that Trino behaves slightly differently than Hive.

Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/BaseTestHiveCoercion.java:273]: Relevant code in Hive

https://github.com/linkedin/linkedin-orc/blob/7b695e34e38f27dd17784341a67326ea0db96903/java/core/src/java/org/apache/orc/impl/ConvertTreeReaderFactory.java#L149-L160

Feel free to add it to the PR description if you consider it noteworthy.

Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/BaseTestHiveCoercion.java:273]: Discussed via Slack and agreed that the difference is related to a peculiarity of the Hive client.

**Review by findinpath — APPROVED** [review]

**Review by marcinsbd — APPROVED** [review]

**Review by krvikash — APPROVED** [review]

**Comment by Praveen2112** [comment]
Updated a few test cases wrt to whitespace

**Review by kokosing — APPROVED** [review]

**Pull request merged**